packages/serverpod_test/lib/src/test_session.dart

```diff
@@ -0,0 +1,142 @@
+    bool? enableLogging,
+  });
+}
+
+/// Internal test session that contains implementation details that should
+/// only be used internally by the test tools.
+class InternalTestSession extends TestSession {
+  final List<InternalTestSession> _allTestSessions;
+  final TestServerpod _testServerpod;
+  AuthenticationOverride? _authenticationOverride;
+  final bool _enableLogging;
+
+  /// The underlying Serverpod session
+  InternalServerpodSession serverpodSession;
+
+  @override
+  Database get db => serverpodSession.db;
+
+  @override
+  Future<AuthenticationInfo?> get authenticationInfo async {
+    return serverpodSession.authenticated;
+  }
+
+  Transaction? _transaction;
+  @override
+  Transaction? get transaction => _transaction;
+
+  set transaction(Transaction? transaction) {
+    _transaction = transaction;
+    serverpodSession.transaction = transaction;
+  }
+
+  /// Creates a new internal test session.
+  InternalTestSession(
+    TestServerpod testServerpod, {
+    AuthenticationOverride? authenticationOverride,
+    InternalTestSession? sessionWithDatabaseConnection,
+    Transaction? transaction,
+    required bool enableLogging,
+    required List<InternalTestSession> allTestSessions,
+    required this.serverpodSession,
+  })  : _allTestSessions = allTestSessions,
+        _authenticationOverride = authenticationOverride,
+        _testServerpod = testServerpod,
+        _enableLogging = enableLogging,
+        _transaction = transaction {
+    _allTestSessions.add(this);
+    _configureServerpodSession(serverpodSession);
+  }
+
+  @override
+  TestSession copyWith({
+    AuthenticationOverride? authentication,
+    bool? enableLogging,
+    String endpoint = '',
+    String method = '',
+  }) {
+    var newServerpodSession = _testServerpod.createSession(
+      enableLogging: enableLogging ?? _enableLogging,
+      transaction: transaction,
+      endpoint: endpoint,
+      method: method,
+    );
+
+    return InternalTestSession(
+      _testServerpod,
+      allTestSessions: _allTestSessions,
+      authenticationOverride: authentication ?? _authenticationOverride,
+      enableLogging: enableLogging ?? _enableLogging,
+      transaction: transaction,
+      serverpodSession: newServerpodSession,
+    );
+  }
+
+  void _configureServerpodSession(InternalServerpodSession session) {
+    var authenticationOverride = _authenticationOverride;
+    if (authenticationOverride is _AuthenticationInfoOverride) {
+      session.updateAuthenticated(authenticationOverride.authenticationInfo);
+    }
+  }
+
+  /// Resets the internal state of the test session
+  /// and recreates the underlying Serverpod session.
+  Future<void> resetState() async {
+    await serverpodSession.close();
+    _authenticationOverride = null;
+    serverpodSession = _testServerpod.createSession(
+      transaction: _transaction,
+      enableLogging: _enableLogging,
+    );
+    _configureServerpodSession(serverpodSession);
+  }
```

Comment:
Question: Where is this used? This will also modify the id of the session! Perhaps that is ok?